### ISSUE TITLE
Implement feature binning

### ISSUE THREAD

**Issue by amaya382** [myui/hivemall#319]
implementation of #294

This implementation is based on `org.apache.hadoop.hive.ql.udf.generic.GenericUDAFPercentileApprox`

# [UDAF] `build_bins(weight, num_of_bins[, auto_shrink])`

### Input

| weight: int&#124;bigint&#124;float&#124;double | num_of_bins: int | [auto_shrink: boolean = false] |
|:--:|:--:|:--:|
| weight | 2 <= | behavior when separations are repeated: T=>skip, F=>exception |

### Output

| quantiles: array\<double> |
|:--:|
| array of separation value |

### Input

*people* table

| name | age | sex |
|:--:|:--:|:--:|
| Jacob | 20 | Male |
| Mason | 22 | Male |
| Sophia | 35 | Female |
| Ethan | 55 | Male |
| Emma | 15 | Female |
| Noah | 46 | Male |
| Isabella | 20 | Female |

```sql
SELECT
  build_bins(age, 3)
FROM
  people
```

### Output

```
[-Infinity,18.333333333333332,30.666666666666657,Infinity]
```

## About `auto_shrink`

There is the possibility quantiles are repeated because of too many `num_of_bins` or too few data.
If `auto_shrink` is true, skip duplicated quantiles. If not, throw an exception.

# [UDF] `feature_binning(features, quantiles_map)/(weight, quantiles)`

#### A

| features: array\<features::string> | quantiles_map: map\<string, array\<double>> |
|:--:|:--:|
| serialized feature | entry:: key: col name, val: quantiles |

#### B

| weight: int&#124;bigint&#124;float&#124;double | quantiles: array\<double> |
|:--:|:--:|
| weight | array of separation value |

#### A

| features: array\<feature::string> |
|:--:|
| serialized and binned features |

#### B

| bin: int |
|:--:|
| categorical value(bin ID) |

#### A

```sql
WITH t AS (
  SELECT
    array_concat(
      categorical_features(
        array("name", "sex"), name, sex),
      quantitative_features(
        array("age"), age)) AS features
  FROM
    people
),
bins AS (
  SELECT map("age", build_bins(age, 3)) AS quantiles_map
  FROM people
)
SELECT feature_binning(features, quantiles_map) AS features FROM t JOIN bins;
```

#### B

```sql
WITH bins AS (
  SELECT build_bins(age, 3) AS quantiles
  FROM people
)
SELECT feature_binning(age, quantiles) AS bin FROM people JOIN bins;
```

#### A

| features: array\<features::string> |
|:--:|
| ["name#Jacob","sex#Male","age:1"] |
| ["name#Mason","sex#Male","age:1"] |
| ["name#Sophia","sex#Female","age:2"] |
| ["name#Ethan","sex#Male","age:2"] |
| ["name#Emma","sex#Female","age:0"] |
| ["name#Noah","sex#Male","age:2"] |
| ["name#Isabella","sex#Female","age:1"] |

#### B
| bin: int |
|:--:|
| 1 |
| 1 |
| 2 |
| 2 |
| 0 |
| 2 |
| 1 |

**Comment by myui** [comment]
- renamed `binning` to `feature_binning`.
- value range should be expressed in two cols instead of an array, say `double::from` (inclusive) and `double::to` (exclusive).

**Comment by myui** [comment]
@amaya382 I think the left most should be -INF and the right most SHOULD be INF. How do you think? It's okey to use range type is fixed to `double`.

**Comment by amaya382** [comment]
@myui I see, fix range type to double. It's easier to implement and understand.

**Comment by amaya382** [comment]
reflected the following comments into implementation plan

> * renamed binning to feature_binning.
> * value range should be expressed in two cols instead of an array, say double::from (inclusive) and double::to (exclusive).
> * I think the left most should be -INF and the right most SHOULD be INF. How do you think? It's okey to use range type is fixed to double.

**Comment by myui** [comment]
```
 +--------+------+---------------------+--------------------+
 | feature | bin |         left        |       right        |
 +--------+------+---------------------+--------------------+
 |   a    | 0    |        -INF         |        1.0         |
 |   a    | 1    |         1.0         |        3.0         |
 |   a    | 2    |         3.0         |       10.0         |
 |   a    | 3    |        10.0         |        INF         |
 |   b    | 0    |        -INF         |       -2.0         |
 |   b    | 1    |        -2.0         |        4.0         |
 |   b    | 2    |         4.0         |        5.0         |
 |   b    | 3    |         5.0         |        INF         |
 +--------+------+---------------------+--------------------+
 ```

```sql
 WITH t1 as (
	 select
		 extract_feature(feature), 
		 feature_binning(extract_weight(features)) as bins -- array<double>
	 from
		 input_table
	 group by
		 extract_feature(feature)
 )
select
	t1.feature,
	t2.bin,
	t2.left,
	t2.right
from	
	t1 LATERAL VIEW explode_bins(bins) t2 as (bin, left, right)
```

**Comment by myui** [comment]
Implemented in #322
